### PR TITLE
dialects: (aie) Add aie dialect

### PR DESCRIPTION
**Pull request by gabrielrodcanal**
Port of the AMD Xilinx AIE dialect for programming the AIEs on the AMD Xilinx Versal FPGA architecture. AIE is a hardened systolic array present in the Versal devices. The dialect describes netlists of AIE components and it can be lowered to the processor's assembly using the vendor's compiler. A description of the original dialect can be found here https://xilinx.github.io/mlir-aie/AIEDialect.html

Other than obvious improvements to the code the reviewers might spot, I would be interested in receiving feedback on the use of `SymbolTable` for `DeviceOp` and the reference to `createObjectFifo` from `ObjectFifoAcquireOp`. My main concern is that the user has to make sure the `createObjectFifo` operation is assigned to a `DeviceOp` before creating an `ObjectFifoAcquireOp` operation that refers to it. Do you see this as an acceptable restriction? If so, should we enforce this through the verifier? Find an example of what I am talking about in https://github.com/xdslproject/mlir-aie/blob/xdsl_tutorials/tutorials/tutorial-4/xdsl-tutorial-4.ipynb. In particular:
```
    object_fifo = aie.createObjectFifo(elem_number, tile14, tile34, i32, [256], "of")
    ...

    @Builder.region
    def region0(builder: Builder):
        ...
        builder.insert(object_fifo)
        ...
        
    device_val = IntegerAttr.from_int_and_width(0, i32)
    device = aie.DeviceOp(device_val, region0)

    acquire_fifo = aie.ObjectFifoAcquireOp(I32Attr(aie.PRODUCE_PORT), I32Attr(1), object_fifo.sym_name, device)
```

This forces the user to add new operations to the `DeviceOp` region after its creation instead of being able to insert all the operations with the `builder`. It looks a bit nasty to me. Example:

```
region0.block.add_op(core14)
```

I would like to request a  review from @mesham, @superlopuh, @math-fehr, @AntonLydike and @PapyChacal.

BTW this doesn't pass Pyright yet, but I'd like to get some feedback before I work on fixing this.

### CODE REVIEW

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:25]: here I would just copy/paste the definition, and also make it private to your file, clients shouldn't need to import TypeVars

Comment (superlopuh) [reply xdsl/dialects/experimental/aie.py:25]: The alternative is to import `AttributeInvT`

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:82]:
```suggestion
class ObjectFIFO(Generic[AttributeInvT], ParametrizedAttribute):
```

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:1]: would be great to have your PR description as a docstring at the top of this file, I'm sure future readers will find it useful, especially external links

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:252]: nit, we've been trying to stick with lowercase names for our python classes, even if that disagrees with the way that's printed in the IR:
```suggestion
    ab: IntegerAttr[i32] = attr_def(IntegerAttr[i32], attr_name="AB")
```

Comment (superlopuh) [reply xdsl/dialects/experimental/aie.py:252]: Also, the type annotation isn't actually necessary, I'm guilty of having started this before realising that only the thing after the `=` is actually needed
```suggestion
    ab = attr_def(IntegerAttr[i32], attr_name="AB")
```

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:411]: similarly here for attribute names and python vs C++ conventions, I'd recommend making python pretty and adding the translation hook thing. I'm not sure if everyone else will agree with this though, definitely a matter of personal preference

**Review by superlopuh — COMMENTED** [review]
The symbol things look good to me. Is it the same as the aie dialect? Would it make sense to add some interop tests to check that this really is a reimplementation of their dialect exactly?

File: xdsl/dialects/experimental/aie.py
Comment (compor) [inline xdsl/dialects/experimental/aie.py:178]: missing `HasParent` trait

File: xdsl/dialects/experimental/aie.py
Comment (compor) [inline xdsl/dialects/experimental/aie.py:151]: Would it make sense for this to be 

```suggestion
        tile: SSAValue | Operation,
```

Comment (compor) [reply xdsl/dialects/experimental/aie.py:151]: Similarly in other places.

File: xdsl/dialects/experimental/aie.py
Comment (compor) [inline xdsl/dialects/experimental/aie.py:153]: Would it make sense to mimick the type annotations from the `MemRefType.from_element_type_and_shape` from `memref`?

**Review by compor — CHANGES_REQUESTED** [review]
This is a massive PR and I'd love to have some test cases that exercise even a few ops.

I left some comments, which I think is a good start, but more iterations to come.

**Review by compor — COMMENTED** [review]
I had a look wrt your `ObjectFifoAcquireOp` and `ObjectFifo` in Xilinx's `mlir-aie`project and to my understanding, there is no reason to expect anything to be "in place" during construction for one to lookup the symbol table.

First, if you want to query `ModuleOp`'s symbol table via the containing `DeviceOp`, there is no guarantee or easy/elegant way to restrict the construction order.
Second, the `ObjectFifoAcquireOp` seems to just store the `FlatSymbolRefAttr` in the TableGen [source][1]. Moreover, it is a flat symbol reference, so no nested lookups seem to be considered.

Thus, I'd do what you suggested and simply store it in `__init__` and use a custom `verify_` to check all the required conditions.
This seems to be what happens in `mlir-aie`, where the `getObjFifoName` is called in the verifier of `ObjectFifoAcquireOp` ([here][2]).

This should simplify construction in your notebook example.
Also, note that you can nest builders in xDSL.

[1]: https://github.com/Xilinx/mlir-aie/blob/76acfeef8faa929b09e9b19d34a9f26834f488e1/include/aie/Dialect/AIE/IR/AIEOps.td#L1608
[2]: https://github.com/Xilinx/mlir-aie/blob/76acfeef8faa929b09e9b19d34a9f26834f488e1/lib/Dialect/AIE/IR/AIEDialect.cpp#L712

File: xdsl/dialects/experimental/aie.py
Comment (gabrielrodcanal) [inline xdsl/dialects/experimental/aie.py:253]: @superlopuh does omiting the type annotation apply to every type of class member? For example for operands do we prefer 
```suggestion
    buffer = operand_def(memref.MemRefType)
```
?

File: xdsl/dialects/experimental/aie.py
Comment (gabrielrodcanal) [reply xdsl/dialects/experimental/aie.py:411]: What would be your suggestion for the name of this attribute?

File: xdsl/dialects/experimental/aie.py
Comment (gabrielrodcanal) [reply xdsl/dialects/experimental/aie.py]: I've tried this but the verifier doesn't seem to be called upon creation of the operation?

Comment (compor) [reply xdsl/dialects/experimental/aie.py]: hmm, weird. If you have that implementation, please push it and we can have a look.

Comment (gabrielrodcanal) [reply xdsl/dialects/experimental/aie.py]: Just pushed it. You can try changing the following in this tutorial https://github.com/xdslproject/mlir-aie/blob/xdsl_tutorials/tutorials/tutorial-4/switchbox/xdsl-tutorial-4-switchbox.ipynb
`connect14 = aie.ConnectOp(aie.WireBundleAttr("DMA"), I32Attr(0), aie.WireBundleAttr("East"), I32Attr(1))`
by
`connect14 = aie.ConnectOp(aie.WireBundleAttr("DMA"), I32Attr(-1), aie.WireBundleAttr("East"), I32Attr(1))`

By changing the `arbiterID` the verifier should be trigger, but that doesn't happen unless `connect14.verify()` is called.

Comment (AntonLydike) [reply xdsl/dialects/experimental/aie.py]: Operation verifiers are not run automatically, they happen at precise points like:
- after input IR parsing is complete
- between passes

If you want to verify IR at other points, you should manually call `verify()` on the module op to verify the whole module, or on any sub-operation you want to verify individually.

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:234]: This is missing a corresponding `parse` function. But also, it feels like this could leverage the format description API instead, like so:
```suggestion
    def print(self, printer: Printer):
        printer.print(
            "<",
            self.sourceBundle.data,
            ": ",
            self.sourceChannel.value.data,
            ", ",
            self.destBundle.data,
            ": ",
            self.destChannel.value.data,
            ">",
        )
        assembly_format = "`<` $sourceBundle `: ` $sourceChannel `, ` $destBundle `: ` destChannel `>`  attr-dict"
```

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:237]: please remove this print

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:338]: looks like this is missing a parse as well

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:364]: what do you mean by this?

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:419]: Something about this feels off to me, I'd expect the textual IR to reflect the value in the attribute. I have a vague recollection of @PapyChacal doing something conceptually related with enums but I might be wrong

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:417]: also missing a `parse` function

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:443]: I would recommend keeping the inits as simple as possible, and letting the client deal with constructing the values.
```suggestion
    def __init__(
        self,
        sym_name: str,
        buffer: memref.MemRefType[Attribute],
    ):
        super().__init__(
            attributes={"sym_name": StringAttr(sym_name)},
            result_types=[buffer],
        )
```

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:901]:
```suggestion
class PutCascade(IRDLOperation):
```
Please make all the classes PascalCase

File: xdsl/dialects/experimental/aie.py
Comment (superlopuh) [inline xdsl/dialects/experimental/aie.py:1128]: nit:
```suggestion
    [WireBundleAttr, ObjectFIFO, ObjectFIFOSubview,],
```
Trailing commas makes it a little easier on git in the future

**Review by superlopuh — CHANGES_REQUESTED** [review]
Looking good! I forget exactly what the agreement that we came up with wrt experimental was. Do we not want any tests? There are a number of operations with a print function but no parse function, so they won't roundtrip, is that a problem? If not, then I guess we could merge once the style comments are addressed

**Comment by gabrielrodcanal** [comment]
> Looking good! I forget exactly what the agreement that we came up with wrt experimental was. Do we not want any tests? There are a number of operations with a print function but no parse function, so they won't roundtrip, is that a problem? If not, then I guess we could merge once the style comments are addressed

If things have not changed since the last conversation we had about experimental we don't want tests. Someone correct me if I'm wrong!

Oh, I think that might be a problem, now that you mention it. I'll address the issue in a following commit before merging.

File: xdsl/dialects/experimental/aie.py
Comment (gabrielrodcanal) [reply xdsl/dialects/experimental/aie.py:234]: I'm getting an error when using the format description API. It's my first time using it, but from what I've seen in other dialects just by removing the print method and having the `assembly_format` definition you provide it should be fine. I get an error during parser though: `'expected variable to refer to an operand, attribute, region, result, or successor'`. It fails with the first variable already, `sourceBundle`.

Comment (superlopuh) [reply xdsl/dialects/experimental/aie.py:234]: Can I have a go at this and push to your PR?

Comment (superlopuh) [reply xdsl/dialects/experimental/aie.py:234]: Ah I spoke with @math-fehr, and it turns out that there isn't support for attributes right now. He says that maybe he can work on it today or tomorrow, not sure if it's worth blocking the PR on this, even if it would make implementing the printers and parsers easier.

Comment (gabrielrodcanal) [reply xdsl/dialects/experimental/aie.py:234]: If it just has to wait one day maybe I'd say it's worth the wait. I have to work on some details of the execution of the code on actual hardware anyways, as it needs some tweaks with respect to the simulator.

Comment (superlopuh) [reply xdsl/dialects/experimental/aie.py:234]: Let's see what @math-fehr says

Comment (superlopuh) [reply xdsl/dialects/experimental/aie.py:234]: #2028

Comment (gabrielrodcanal) [reply xdsl/dialects/experimental/aie.py:234]: Just merged main with this branch and I keep getting the same error.

Comment (superlopuh) [reply xdsl/dialects/experimental/aie.py:234]: Can you please push the broken version?

File: xdsl/dialects/experimental/aie.py
Comment (PapyChacal) [reply xdsl/dialects/experimental/aie.py:417]: Completely missed that ping!
I implemented EnumAttribute, but it wouldn't help for printing a keyword-prefixed integer?
The declarative assembly format soon should have support for that.
That's the job of optional groups, that would be written 
```
( `xcvc1902` $device^ )?
```
And attribute directives and optional groups are both open on separate PRs :slightly_smiling_face:

**Comment by tobiasgrosser** [comment]
> > Looking good! I forget exactly what the agreement that we came up with wrt experimental was. Do we not want any tests? There are a number of operations with a print function but no parse function, so they won't roundtrip, is that a problem? If not, then I guess we could merge once the style comments are addressed
> 
> If things have not changed since the last conversation we had about experimental we don't want tests. Someone correct me if I'm wrong!
> 
> Oh, I think that might be a problem, now that you mention it. I'll address the issue in a following commit before merging.

We should not unnecessarily slow down things in experimental. On the other side, I really wonder if test cases indeed slow down development so much. If they have a low cost, I would invite test cases but would maybe not insist on them,

File: xdsl/dialects/experimental/aie.py
Comment (gabrielrodcanal) [reply xdsl/dialects/experimental/aie.py:234]: @superlopuh sorry, didn't get a notification of this! The broken version is already pushed.

Comment (superlopuh) [reply xdsl/dialects/experimental/aie.py:234]: It looks like all the tests pass, there's no error. Could you please push the filecheck also?

Comment (gabrielrodcanal) [reply xdsl/dialects/experimental/aie.py:234]: Sorry, I didn't commit the broken version in a "broken" state, but with the custom print method and I thought you would replace it with the API line on your end as I was doing! Just comitted the broken version and now you can see how a test fails and gives the same error I get.

Comment (superlopuh) [reply xdsl/dialects/experimental/aie.py:236]: Thank you, I merged main and fixed a few issues. I also added a super basic filecheck test. It feels like it would be worth the effort to add the other ops for which you'd like to have custom syntax to the filecheck, just to make sure that they roundtrip ok

**Review by superlopuh — APPROVED** [review]
Great! I think it might be worth having a go at replacing all the `print` without a corresponding `parse` with the assembly_format, but don't want to slow you down from now :)

**Comment by gabrielrodcanal** [comment]
> Great! I think it might be worth having a go at replacing all the `print` without a corresponding `parse` with the assembly_format, but don't want to slow you down from now :)

Yes! I'm working on it :)

File: xdsl/dialects/experimental/aie.py
Comment (gabrielrodcanal) [reply xdsl/dialects/experimental/aie.py]: That statement makes use of an EnumAttribute for the type of device. I'm afraid the declarative assembly format is not an option here because this operation has a region. Unless this capability has been added over the last week.

**Review by compor — COMMENTED** [review]
I understand that this goes to `experimental` but I think it might worth revisting how we do things in general regarding this.
What would constitute "graduating" from `experimental`?

Considering how long this took to "finalize", I am not sure writing tests would slow down development that much.

On the flip side, we are called to review a massive PR and regain context after weeks of giving an initial review.

Considering this, I don't feel confident approving this, but I don't want to block it.
So, please go ahead without my explicit approval.

**Review by mesham — APPROVED** [review]
I think this looks good - it's adding the AMD AIE dialect into xDSL via experimental and as Gabriel has actioned the specific suggestions and comments on this I think it's good to get it checked in and usable more widely. There are some actions to write tests around this in the future, but many things in experimental lack tests (and indeed the point of experimental is to get less mature things in and available sooner) so it won't be unique in this regard. As experimental is explicitly "use at your own risk, this is less mature", then I think this is fine and nice to expand the available dialects.

File: xdsl/irdl/declarative_assembly_format_parser.py
Comment (tobiasgrosser) [inline xdsl/irdl/declarative_assembly_format_parser.py:323]: I would suggest submitting this through a separate PR, as this is indeed not in the experimental folder or taking this out of this PR.

**Review by tobiasgrosser — APPROVED** [review]
I am fine adding this for now. Then we can go with smaller PRs that incrementally improve what we have.

File: xdsl/irdl/declarative_assembly_format_parser.py
Comment (gabrielrodcanal) [reply xdsl/irdl/declarative_assembly_format_parser.py:323]: I think this is a change from main, if I click on 'View file' it takes me to the merge @superlopuh did from main, so I guess there's no need for a new PR?

Comment (tobiasgrosser) [reply xdsl/irdl/declarative_assembly_format_parser.py:329]: Can you merge latest main and make sure this diff disappears?

**Comment by tobiasgrosser** [comment]
OK, that is good to go for me. I personally would add tests but that's up to you, I guess.

**Comment by gabrielrodcanal** [comment]
> OK, that is good to go for me. I personally would add tests but that's up to you, I guess.

I will add tests in the near future, once I'm done with some deadlines!

**Pull request merged**
